File: api/entities/puzzle.go
package entities

import (
	"time"

	"github.com/RagOfJoes/puzzlely/internal/validate"
	"github.com/google/uuid"
)

var _ Entity = (*Puzzle)(nil)

// PuzzleDifficulty defines the difficulty options for a puzzle
type PuzzleDifficulty string

const (
	Easy   PuzzleDifficulty = "Easy"
	Medium PuzzleDifficulty = "Medium"
	Hard   PuzzleDifficulty = "Hard"
)

// Puzzle defines a puzzle
type Puzzle struct {
	Base

	// Name defines the name of the puzzle
	Name string `json:"name" validate:"required,notblank,alphanumspace,min=1,max=64"`
	// Description can be used as hints, red herrings, etc.
	Description string `json:"description,omitempty" validate:"printasciiextra,max=512"`
	// Difficulty of the connection between blocks
	Difficulty PuzzleDifficulty `json:"difficulty" validate:"required,oneof='Easy' 'Medium' 'Hard'"`
	// MaxAttempts limits how a user can configure their game when playing the puzzle
	//
	// If 0, users will be able to customize their game's maxAttempt config
	MaxAttempts uint16 `json:"maxAttempts" validate:"min=0,max=999"`
	// TimeAllowed limits how a user can configure their game when playing the puzzle
	//
	// If 0, users will be able to customize their game's timeAllowed config
	TimeAllowed uint32 `json:"timeAllowed" validate:"min=0,max=3599000"`
	// LikedAt defines when or if a user has like the puzzle
	LikedAt *time.Time `json:"likedAt"`
	// NumOfLikes defines the number of likes of the puzzle
	NumOfLikes uint `json:"numOfLikes"`
	// Groups that belong to this puzzle
	Groups []PuzzleGroup `json:"groups" validate:"required,len=4,dive"`
	// CreatedBy defines the user that created the puzzle
	CreatedBy User `json:"createdBy" validate:"required"`
}

// NewPuzzle creates a new puzzle for a given user
func NewPuzzle(name, description string, difficulty PuzzleDifficulty, maxAttempts uint16, timeAllowed uint32, groups []PuzzleGroup, createdBy User) Puzzle {
	return Puzzle{
		Base: Base{
			ID:        uuid.New(),
			CreatedAt: time.Now(),
		},

		Name:        name,
		Description: description,
		Difficulty:  difficulty,
		MaxAttempts: maxAttempts,
		TimeAllowed: timeAllowed,
		Groups:      groups,
		CreatedBy:   createdBy,
	}
}

func (p *Puzzle) Validate() error {
	return validate.Check(p)
}

File: web/src/types/puzzle.ts
import type { User } from "./user";

/**
 * Base
 */

export type Block = {
  /**
   * Unique identifier
   */
  id: string;
  /**
   * Unique identifier of the group that this block belongs to
   */
  groupID: string;
  /**
   * Text that will be displayed on the ui
   */
  value: string;
};
export type Group = {
  /**
   * Unique identifier
   */
  id: string;
  /**
   * Array of string fragments that can be used to compare user input with intended answer
   */
  answers: string[];
  /**
   * Description for the group that can explain the linkage, fun facts, etc.
   */
  description: string;
  /**
   * Blocks that belongs to this group
   */
  blocks: Block[];
};
export type Puzzle = {
  /**
   * Unique identifier
   */
  id: string;
  /**
   * Name of the puzzle
   */
  name: string;
  /**
   * Difficulty of the connection between blocks
   */
  difficulty: "Easy" | "Medium" | "Hard";
  /**
   * Description can be used as hints, red herrings, etc.
   */
  description?: string;
  /**
   * Restrict what `config.maxAttempts` setting a user can configure
   */
  maxAttempts: number;
  /**
   * Restrict what `config.timeAllowed` setting a user can configure
   */
  timeAllowed: number;
  /**
   * Number of likes
   */
  numOfLikes: number;
  /**
   * Whether user has liked Puzzle
   */
  likedAt?: Date | null;
  /**
   * When the puzzle was created
   */
  createdAt: Date;
  /**
   * If and when the puzzle was update
   */
  updatedAt?: Date;
  /**
   * Group of blocks. For now, this is limited to 4.
   *
   * TODO: Allow this to be customizable
   */
  groups: Group[];
  /**
   * Who created the puzzle
   */
  createdBy: User;
};

/**
 * Payloads
 */

export type PuzzleCreatePayload = Pick<
  Puzzle,
  "name" | "description" | "difficulty" | "maxAttempts" | "timeAllowed"
> & {
  groups: (Pick<Group, "answers" | "description"> & {
    blocks: { value: Block["value"] }[];
  })[];
};
export type PuzzleUpdatePayload = Pick<
  Puzzle,
  "name" | "description" | "difficulty"
> & {
  groups: Pick<Group, "id" | "description">[];
};

/**
 * Cursor pagination
 */

export type PuzzleNode = Omit<Puzzle, "groups">;
export type PuzzleEdge = {
  cursor: string;
  node: PuzzleNode;
};
export type PuzzleConnection = {
  edges: PuzzleEdge[];
  pageInfo: {
    cursor: string;
    hasNextPage: boolean;
  };
};

/**
 * Misc.
 */

export type PuzzleFilters =
  | "customizable_attempts"
  | "customizable_time"
  | "difficulty"
  | "num_of_likes";
export type PuzzleLike = {
  id: string;
  active: boolean;
  createdAt: Date;
  updatedAt?: Date | null;
};
